### PR TITLE
Suppress previous route transition if current route is fullscreenDialog

### LINKED ISSUE #159289: Fullscreen route transition broken on 3.27.0-0.2.pre beta
### Steps to reproduce

* Check the video
* Run the provided sample with Flutter 3.27.0-0.2.pre and verify the results
* Run the provided sample with an earlier stable Flutter version which works as expected (I tested with 3.22.3 and 3.24.0)

I think the cause is likely https://github.com/flutter/flutter/pull/150031 and created the issue as requested there.

### Expected results

The underlying route should not be transitioned out when a fullscreen route is pushed on top.

### Actual results

The underlying route is transitioned out to the left side while the fullscreen route is pushed from bottom up. This leads to a broken transition animation.

### Code sample

https://github.com/kuhnroyal/flutter_route_transition

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]
[Screen_recording_20241121_210951.webm](https://github.com/user-attachments/assets/d301238d-6615-42a7-b60a-611c61136d88)

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel beta, 3.27.0-0.2.pre, on macOS 14.7.1 23H222 darwin-arm64, locale de-DE)
    • Flutter version 3.27.0-0.2.pre on channel beta at /Users/########r/fvm/versions/3.27.0-0.2.pre@beta
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fc011960a2 (vor 7 Tagen), 2024-11-14 12:19:18 -0800
    • Engine revision 397deba30f
    • Dart version 3.6.0 (build 3.6.0-334.4.beta)
    • DevTools version 2.40.1

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/#######/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = /Users/#######/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/#######/Library/Android/sdk
    • Java binary at: /Users/########/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 16.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16B40
    ✗ CocoaPods installed but not working.
        You appear to have CocoaPods installed but it is not working.
        This can happen if the version of Ruby that CocoaPods was installed with is different from the one being used to invoke it.
        This can usually be fixed by re-installing CocoaPods.
      For re-installation instructions, see https://guides.cocoapods.org/using/getting-started.html#installation

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.2)
    • Android Studio at /Users/########/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)

[✓] IntelliJ IDEA Ultimate Edition (version 2024.2.4)
    • IntelliJ at /Users/#########/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 82.1.3
    • Dart plugin version 242.22855.32

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile)     • emulator-5554         • android-arm64  • Android 15 (API 35) (emulator)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.7.1 23H222 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.7.1 23H222 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 131.0.6778.71
    ! Error: Browsing on the local area network for iPhone 14 Pro Max von ########. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

### PR DESCRIPTION
**Pull request by MitchellGoodwin**
Fixes #159289

Cupertino and Material routes were looking to see if the next route was a full screen dialog route, and not transitioning accordingly. However with the updates to allow [mixed transitions in an app](https://github.com/flutter/flutter/pull/150031), any route will try and transition if a delegatedTransition is available. This PR makes it so that Cupertino and Material routes that are fullscreen dialogs will use `canTransitionFrom` to tell the previous route not to transition.

Before:
[388677067-d301238d-6615-42a7-b60a-611c61136d88.webm](https://github.com/user-attachments/assets/eea99ee9-0bc8-4981-a950-08f99a7fdb3e)

After:

https://github.com/user-attachments/assets/7deb5143-cd67-4696-95ec-9d7df329dfa4

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/cupertino/route_test.dart [hunk 1]
```diff
@@ -2691,6 +2691,54 @@ void main() {
...
   });
 
+  testWidgets('fullscreen routes do not transition previous route', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      CupertinoApp(
+        initialRoute: '/',
+        onGenerateRoute: (RouteSettings settings) {
+          if (settings.name == '/') {
+            return PageRouteBuilder<void>(
+              pageBuilder: (_, __, ___) {
```
Comment (justinmc) [inline packages/flutter/test/cupertino/route_test.dart:2701]: Nit: [Avoid anonymous parameter names](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-anonymous-parameter-names), though I know we don't always follow this in tests...

Also a few other instances below.

File: packages/flutter/test/cupertino/route_test.dart [hunk 2]
```diff
@@ -2691,6 +2691,54 @@ void main() {
...
+    );
+
+    expect(find.text('Page 1'), findsOneWidget);
+    expect(find.text('Page 2'), findsNothing);
+
+    final double pageTitleDX = tester.getTopLeft(find.text('Page 1')).dx;
+
+    tester.state<NavigatorState>(find.byType(Navigator)).pushNamed('/next');
+    await tester.pump();
+    await tester.pump(const Duration(milliseconds: 100));
```
Comment (justinmc) [inline packages/flutter/test/cupertino/route_test.dart:2733]: Why are two `pump`s needed here? Sorry if this is just copying another test that I already asked about, this looks familiar but I can't remember.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Thanks for the quick fix for this.

File: packages/flutter/test/cupertino/route_test.dart (hunk 2 above)
Comment (MitchellGoodwin) [reply packages/flutter/test/cupertino/route_test.dart:2733]: An extra pump before a pump with a duration is necessary to push the route and begin the animation.

**Comment by RepliedSage11** [comment]
@justinmc Hi, is there a timeline to get this as a hot fix for 3.27 stable?

**Pull request merged**